servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java

```diff
@@ -226,8 +226,27 @@ private DnsDiscoveryObserver newDiscoveryObserver(final String address) {
     public Publisher<Collection<ServiceDiscovererEvent<InetAddress>>> dnsQuery(final String address) {
         requireNonNull(address);
         return defer(() -> {
-            ARecordPublisher pub = new ARecordPublisher(address, newDiscoveryObserver(address));
-            return inactiveEventsOnError ? recoverWithInactiveEvents(pub, false) : pub;
+            final DnsDiscoveryObserver discoveryObserver = newDiscoveryObserver(address);
+            ARecordPublisher pub = new ARecordPublisher(address, discoveryObserver);
+            Publisher<? extends Collection<ServiceDiscovererEvent<InetAddress>>> events = inactiveEventsOnError ?
+                    recoverWithInactiveEvents(pub, false) :
+                    pub;
+            return events
+                    .afterCancel(() -> {
+                        if (discoveryObserver != null) {
+                            discoveryObserver.discoveryCanceled();
```

Comment:
This is what I was talking about previously :) without "catch-all" observer wrapper it's so easy to miss a try-catch for a new callback. 
`DnsServiceDiscovererObserver` was the first observer we implemented (before `TransportObserver`). Here, the need for "catch-all" wrapper was less obvious because the entire scope is within one file. Will be nice to make it consistent for `DnsServiceDiscovererObserver` too, but not necessary for this PR as it will expand the scope. Let's add try-catches for now and we can introduce "catch-all" wrapper in a follow-up.